sdrplay: init at 3.07.1

pmenke-de opened the pull request:
This adds support for software defined radio (SDR) devices by SDRPlay.
SDRPlay provides an unfree binary library and api-service as well
as a MIT licensed adapter library for SoapySDR for integration
with many popular SDR applications.

###### Motivation for this change

I own a RSP1A SDR, set it up under NixOS and would like to share the necessary expressions.

###### Notes

 - This is my first new package and my first module for NixOS. So my experience on "how it should be done" is limited :smile: I think it would good to have this reviewed by some experienced. I'm happy to learn and apply recommended changes.
 - I'm not 100% sure about the sdrplay lib/api-service license:
   - As it's packaged within some [run-binary](https://www.sdrplay.com/software/SDRplay_RSP_API-Linux-3.07.1.run) I posted a copy of it [here](https://gist.github.com/pmenke-de/48009d91317a449042867f4def544e46).
   - It mentions, that »You may reproduce and distribute copies of the Product in any medium, without modifications, in Object form [...]«, but under conditions that (as far as I understand) the license text must somehow be "given" to the user.
   - As I'm not qualified to correctly interpret this, I set the `meta.license` to `unfree`. Maybe a case could be made for `unfreeRedistributable`.
 - [soapysdr-sdrplay3](https://github.com/pothosware/SoapySDRPlay3) doesn't have any releases or tags on github, what is the appropriate `version` attribute in this case? For now i've arbitrarily chosen `0.0.1`. 
 - @lasandell in #60337 you mentioned »working on support for SDRPlay«. Maybe this PR is interesting / relevant to you as well.


###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Manually tested SDR functionality with Gqrx and RSP1A hardware.
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

lasandell commented:
I have this working for my personal use, but it was for libsdrplay v2, and I never submitted it to nixpkgs (see https://github.com/lasandell/nixpkgs/commit/2696671c1c79633850a27dd174145886f17a87f0). We could have both as packages (libsdrplay2, libsdrplay3, soapsdrplay2, soapsdrplay3, etc) since they are architecturally different. Also because version 2 may still be needed for osmosdr / gqrx support which I got working in this throwaway commit: https://github.com/lasandell/nixpkgs/commit/6aa4915d08bb1db116761ae19ae48aba6497479b#diff-32e15e5b491dfc7cf46443c8531f05e0fe32f68a9b57f37694ce9187e8fe61c7

As far as the version, I think sometimes people use timestamps for versions without a tag such as 20210425 but I'm not sure of the best convention. I'm also not sure whether libsdrplay or sdrplay is the best name, here's Repology on what other distros are doing: https://repology.org/projects/?search=sdrplay. 

One problem with the  `lib.optionals config.allowUnfree [ soapysdrplay ]` approach is that it will cause soapysdrplay and the packages that depend on it to be built from source when someone has the `allowUnfree` flag set, so I don't think this is acceptable (Hydra won't build/cache anything with `unfree` or `unfreeRedistributable` licenses). If someone wants to use this package, we may need them to explicitly opt into it using an overlay. I do this in my nixos config as follows:

```
nixpkgs.overlays = [ (self: super: {
  soapysdr-with-plugins = super.soapysdr-with-plugins.override { extraPackages = with self; [ soapysdrplay ]; };
```

Those are my thoughts for now. I may try to test this out but probably won't have time until later this week.

pmenke-de commented:
> Also because version 2 may still be needed for osmosdr / gqrx support

Actually I'm using this PR with gqrx right now. So the osmosdr / gqrx track seems to work with libsdrplay3 too.

> One problem with the lib.optionals config.allowUnfree [ soapysdrplay ] approach

Yeah, I thought, that could be problematic but wasn't sure on the convention here. The problem i had with overriding `soapysdr-with-plugins` was, that i found no other way to extend the `extraModules` list, that's already there in `all-packages.nix`, instead of replacing it. But also maybe that no as much of an issue, as most people probably have just one or two SDRs and don't need all plugins.
I also thought about putting sth. like an `withSdrplay ? false` argument into `soapysdr-with-plugins`, which one would override to `true` to include the unfree plugin alongside the other ones, but was unable to implement it - i.d.k. if it's even possible.

> We could have both as packages (libsdrplay2, libsdrplay3, soapsdrplay2, soapsdrplay3, etc)

If there is a need / use-case for v2 too, then yes, we could implemented that too and have the major version in the package name. Would you prefer this as part of this same PR or separately? I would give it a try to implement it.

> I may try to test this out but probably won't have time until later this week.

No pressure. It's a hobby project for me too :smile: 
Thanks for your quick response.

lasandell commented:
> Actually I'm using this PR with gqrx right now. So the osmosdr / gqrx track seems to work with libsdrplay3 too.

> If there is a need / use-case for v2 too, then yes, we could implemented that too and have the major version in the package name. Would you prefer this as part of this same PR or separately?

If nothing depends on the old version anymore then I would favor leaving the old package out and just having a single sdrplay / soapysdrplay package.

> I also thought about putting sth. like an withSdrplay ? false argument into soapysdr-with-plugins, which one would override to true to include the unfree plugin alongside the other ones, but was unable to implement it - i.d.k. if it's even possible.

That sounds like a reasonable solution. Maybe call it `sdrplaySupport` since that naming convention is used [elsewhere  in nixpkgs](https://github.com/NixOS/nixpkgs/search?p=3&q=support).

> Thanks for your quick response.

No problem, glad to provide input!

markuskowa commented on nixos/modules/services/misc/sdrplay.nix:
```diff
@@ -0,0 +1,15 @@
+{ config, lib, pkgs, sdrplay, ... }:
+with lib;
+{
+  options.services.sdrplay_api = {
```
```suggestion
  options.services.sdrplayApi = {
```

pmenke-de replied:
I changed the name according to your suggestion.

markuskowa commented on pkgs/applications/radio/sdrplay/default.nix:
```diff
@@ -0,0 +1,46 @@
+  arch = if stdenv.isx86_64  then "x86_64"
+    else if stdenv.isi686    then "i686"
+    else throw "unsupported architecture";
+in
```
```suggestion
in stdenv.mkDerivation rec {
```

pmenke-de replied:
Done.

markuskowa commented on pkgs/applications/radio/sdrplay/default.nix:
```diff
@@ -0,0 +1,46 @@
+
+  sourceRoot = "source";
+
+  buildPhase = "";
```
```suggestion
  dontBuild = true;
```
(I think that should work and it is cleaner)

pmenke-de replied:
Yep. Looks good and cleaner. Thanks for the hint.

markuskowa commented on pkgs/applications/radio/soapysdrplay/default.nix:
```diff
@@ -0,0 +1,31 @@
+    description = "Soapy SDR module for SDRPlay";
+    homepage = "https://github.com/pothosware/SoapySDRPlay3";
+    license = licenses.mit;
+    maintainers = [ pmenke ];
```
```suggestion
    maintainers = [ maintainers.pmenke ];
```
I think this causes the ofBorg check to fail.

pmenke-de replied:
Ok. I updated it according to your suggestion.

markuskowa commented on pkgs/applications/radio/sdrplay/default.nix:
```diff
@@ -0,0 +1,46 @@
+  version = "3.07.1";
+
+  # soapysdrplay also needs to full so name
+  soName = "libsdrplay_api.so." + lib.concatStringsSep "." (lib.take 2 (builtins.splitVersion version));
```
```suggestion
  passthru.soName = "libsdrplay_api.so." + lib.concatStringsSep "." (lib.take 2 (builtins.splitVersion version));
```
You are also using it in the soapysdr module. I think this should be a `passthru` property.

pmenke-de replied:
Taking some inspiration from the sdrplay2 implementation of @lasandell i was able elide this property completely :smile: 
I have yet to understand how this pkg-config automagic works though.

markuskowa commented on pkgs/applications/radio/sdrplay/default.nix:
```diff
@@ -0,0 +1,46 @@
+  meta = with lib; {
+    description = "SDRPlay API";
+    homepage = "https://www.sdrplay.com/downloads/";
+    license = licenses.unfree;
```
Is there a link to the license agreement/conditions that we could add here?

pmenke-de replied:
I didn't find any direct license page / link, but a [documentation PDF](https://www.sdrplay.com/docs/SDRplay_API_Specification_v3.07.pdf). I've now included that in in the `longDescription`. Is that the right place for it?

markuskowa reviewed (changes requested):
Just a bunch of small comments. Looks good otherwise.

aanderse commented on nixos/modules/services/misc/sdrplay.nix:
```diff
@@ -0,0 +1,15 @@
+  };
+
+  config = mkIf config.services.sdrplay_api.enable {
+    systemd.services.sdrplay_api = {
```
Does the package contain any `systemd` units, or explain how the software should be run? As is this still run the software as the `root` user - something we're generally not fond of in NixOS, especially when it is proprietary software.

pmenke-de replied:
There is indeed some `systemd` unit inside the `.run` package from SDRplay. Interestingly it explicitly says `User=root`, but it seems the service runs just fine without any privileges. So i changed it to `DynamicUser = true`, which is like "dynamic user with no permissions" it I understand that mechanism correctly.

aanderse replied:
Looks awesome! Bonus points awarded for use of `DynamicUser` 🎉

aanderse commented on nixos/modules/services/misc/sdrplay.nix:
```diff
@@ -0,0 +1,15 @@
+  config = mkIf config.services.sdrplay_api.enable {
+    systemd.services.sdrplay_api = {
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig.ExecStart = "${sdrplay}/bin/sdrplay_apiService";
```
```suggestion
      serviceConfig.ExecStart = "${pkgs.sdrplay}/bin/sdrplay_apiServipkgs
```

pmenke-de replied:
I guess the trailing `pkgs` instead of a `"` in your line was accidental?
I've applied the change of accessing `sdrplay` through `pkgs.sdrplay` and removed the function argument at the top.

aanderse commented on nixos/modules/services/misc/sdrplay.nix:
```diff
@@ -0,0 +1,15 @@
+{ config, lib, pkgs, sdrplay, ... }:
```
```suggestion
{ config, lib, pkgs, ... }:
```

aanderse reviewed (changes requested):
I left a few additional comments on the module. Any information you can provide on running the software would be useful. Thanks!

pmenke-de committed:
sdrplay: init at 3.07.1

this adds support for software defined radio (SDR) devices by SDRplay.
SDRplay provides an unfree binary library and api-service as well
as a MIT licensed adapter library for SoapySDR for integration
with many popular SDR applications.

pmenke-de commented:
Thank you all for all of your input. 
I applied your suggestions and refined the implementation with some inspiration i took from the sdrplay2 implementation of lasandell@2696671

pmenke-de commented:
> Any information you can provide on running the software would be useful

Do you mean how to run / test the software?

I guess you would need some of SDRplays software defined radios for that. I've got one of their [RSP1A](https://www.sdrplay.com/rsp1a/) receivers. This API service seems to manage/proxy access to the device (which is connected via USB) and is in turn called by their API library. They also provide a SoapySDR plugin implementation, which uses this API. So to actually do sth. with the device one could for example use the `gqrx` SDR GUI application; or for a trivial use-case just call `SoapySDRUtil --find`.

I don't exactly know the scope of the NixOS documentation or how exactly it is maintained. Would it be desirable to have a section in there, describing how the you need to configure NixOS to actually support SDRplay? Sth. like, but more fleshed out:

 - put `services.sdrplayApi.enable = true;` in your config 
 - add `soapysdr-with-plugins.override { sdrplaySupport = true; }` as an overlay.

markuskowa commented on pkgs/applications/radio/sdrplay/default.nix:
```diff
@@ -0,0 +1,51 @@
+      https://www.sdrplay.com/docs/SDRplay_API_Specification_v${lib.concatStringsSep "." (lib.take 2 (builtins.splitVersion version))}.pdf
+    '';
+    homepage = "https://www.sdrplay.com/downloads/";
+    license = licenses.unfree;
```
The license in that document is not very specific/clear. It actually reads like a BSD 3-clause license (which would be a free license and should allow for access to the source code). I think it is OK now as it is. We are on the safe side with `unfree`.

pmenke-de replied:
Now that you mention it. That license look rather different compared to the one that ships in the `.run` binary. I should have checked that more carefully. I posted the shipped one in a [gist](https://gist.github.com/pmenke-de/48009d91317a449042867f4def544e46) as it seems you otherwise need to download and extract that binary to obtain it.

aanderse commented:
If the software isn't obvious you can include some details in the `enable` option description, etc...

The module looks good to me now :+1:

pmenke-de commented:
> If the software isn't obvious you can include some details in the enable option description, etc...

Ok. I added a short hint to the enable option description.

pmenke-de commented:
ofBorg seems still to be unhappy though. Is there something left i need to do, or is it just, that i'm »not a trusted user«?

lasandell commented:
I tested with my SDRPlay and it does work with CubicSDR and GQRX. 

It looks like the OfBorg error might be due to it not liking something about the use makeOverridable, possibly because it returns a function taking a attribute set rather than a derivation like callPackage would. I'm not sure of the fix for that.

markuskowa commented on pkgs/top-level/all-packages.nix:
```diff
@@ -17896,18 +17898,21 @@ in
-  };
+  soapysdrplay = callPackage ../applications/radio/soapysdrplay { };
+
+  soapysdr-with-plugins = makeOverridable ({ sdrplaySupport ? false }:
```
Using `makeOverridable` is not a good choice here. It masks all other inputs for possible overrides (besides failing to evaluate).
Since you need an override anyway to activate the plugin you could use `extraPackages` directly:
```
soapysdr-with-plugins = super.soapysdr-with-plugins.override {
  extraPackages = with super; [
    soapysdrplay
    <other plugins of your choice >
  ];
};
```

lasandell replied:
Only problem with that is you have to list out all the default packages if you still want them. I think there might be a way to make it work correctly with `makeOverridable` - I'll try to look into that later today.

lasandell replied:
I wasn't able to get that to work. You may just have to scrap the `sdrplaySupport` flag and force the user to list everything they want in `extraPackages` like @markuskowa says. Or maybe you could have a `soapysdr-with-sdrplay` entry in all-packages.nix that included all the default packages plus soapysdrplay. There might be a more elegant solution but I'm not able to come up with anything right now.

pmenke-de replied:
Thanks for giving it a try too.
I would then indeed just scrap the idea of having this `sdrSupport` flag. I updated the implementation accordingly and changed documentation on the sdrplayApi module to suggest an override of the `extraPackages` argument.

markuskowa replied:
@lasandell I agree that it is a disadvantage that you now have to list all plugins explicitly in the override.
However, on an individual setup you may only need a few of them anyway. On an embedded system you might even want to keep it small.

pmenke-de commented:
Thank you all for reviewing and enhancing this PR.
I must say it has been a really nice experience to work with and learn from you <3
